Add ORMLite comparison

kenkyee opened the pull request:
Pretty big checkin because I had to change everything to work w/ Collections instead of Lists to make ORMLite integration easier.

kenkyee committed:
Don't do DB benchmarks in UI thread and put the results in the UI instead of in the log

kenkyee committed:
Merge branch 'feature/ken/dbflow2.0' into feature/ken/nonlockedui

kenkyee committed:
Added OrmLite support; switched to Collections instead of Lists for compatibility

PerfectCarl commented:
Nice work. Do you have any idea where ormlite ranks? Ormlite author was conceding on the forums that ormlite tends to create a lit of intermediate objects that trigger the gc more often than say greendao taxing the performance ultimately

kenkyee commented:
Looks like roughly 3rd place (after DBFlow and GreenDAO).
I should actually add charting to this app and include a screenshot of the graph for this repo...I'll see if I can do that today.
What hurts ORMLite is you can't give it a nested collection (e.g., address book has contacts but each contact can have a collection of different addresses) and have it just insert everything.  You have to make separate calls for each subcollection after you add the parent collection.  This is supposed to be fixed in ORMLite 4.49 (current version is 4.48) but I'm not sure how far they're going to go w/ nested collections.

kenkyee committed:
Graph results and add resulting charts to the readme.

agrosner commented on README.md:
```diff
@@ -1,4 +1,20 @@
-**Note** more to come soon about these with results and methods.
+## Benchmark Description
+
+There are two benchmarks.  The Simple trial uses a flat schema for an address book so each row is composed of name, address, city, state, and phone columns.  The Complex trial is hierarchical and has support for multiple address books where each address book has contacts and addresses.
```
#req we should separate each trial and diagram the schema in a visual way or by listing each table and relationship. (UML?)

agrosner commented on app/build.gradle:
```diff
@@ -18,12 +18,20 @@ android {
-    apt 'com.raizlabs.android:DBFlow-Compiler:1.4.5'
-    mod 'com.raizlabs.android:{DBFlow, DBFlow-Core}:1.4.5'
+    apt 'com.raizlabs.android:DBFlow-Compiler:2.0.0'
+    compile "com.raizlabs.android:DBFlow-Core:2.0.0"
```
#req not sure why this changed. Just use:
`mod 'com.raizlabs.android:{DBFlow, DBFlow-Core}:2.0.0`

agrosner commented on app/src/main/java/com/raizlabs/android/databasecomparison/MainActivity.java:
```diff
@@ -1,58 +1,213 @@
+    public void runComplexTrial(View v) {
+        setBusyUI(true, getResources().getString(R.string.complex));
+        resetChart();
+        final MainActivity mainActivity = this;
```
#req may leak activity here. I think we should utilize some kind of `AsyncTask` outside of this activity that contains a callback when the result finishes and the `Activity` can register a listener on it.

kenkyee replied:
are you sure this pattern leaks an activity?  It's not like the AsyncTask pattern w/ an implicit Activity reference that leaks a ref.  The thread should release the ref to the Activity when it finishes (I think?)....

agrosner replied:
yes, The tests take some time and if you leave the activity while it runs and reopen the app, the old `Activity` will still remain. I think. Its not a huge deal

agrosner commented on app/src/main/java/com/raizlabs/android/databasecomparison/interfaces/IAddressBook.java:
```diff
@@ -0,0 +1,21 @@
+import java.util.Collection;
+
+/**
+ * Description:
```
#req write description here.

agrosner commented on app/src/main/java/com/raizlabs/android/databasecomparison/ormlite/DatabaseHelper.java:
```diff
@@ -0,0 +1,138 @@
+
+        // here we try inserting data in the on-create as a test
+        /*
+        RuntimeExceptionDao<SimpleData, Integer> dao = getSimpleDataDao();
```
#req remove commented out code.

agrosner commented on app/src/main/res/layout/activity_main.xml:
```diff
@@ -9,15 +9,49 @@
-        android:text="Complex Trial"
-        android:id="@+id/complex"
-        android:layout_width="wrap_content"
+    <LinearLayout
```
#req use the studio autoformat here. It will arrange the tags appropriately.

agrosner commented on README.md:
```diff
@@ -1,4 +1,20 @@
+
+And these are the results for the Complex trial:
+
+![Alt text](images/complextrial.png "Disk Usage")
```
#req the graph for loading appears too small to see. Can you adjust its max value or for every graph determine the max value and use that as the top Y-axis?

kenkyee replied:
I think the real solution is to get rid of the really slow ones.  I tried tuning Sprinkles because it looked like there was bug in that it the transaction wasn't being passed to the save call but it made no difference :-P
The max value is the max value + height of the bar's value text.  Looks like it's calculating it right IMHO.
The graph is only usable in landscape too...not enough horizontal space in portrait :-(

agrosner commented on ormlitegenerator/build.gradle:
```diff
@@ -0,0 +1,67 @@
+import org.apache.tools.ant.taskdefs.condition.Os
```
#? ORMlite isn't _really_ that light. It appears to require an insane amount of configuration.....

kenkyee replied:
no different than GreenDAO...it essentially requires a compilation step to get rid of the reflection

kenkyee committed:
address some of the PR comments

kenkyee committed:
add UML diagrams from SchemaSpy for DB (had to use DBFLow's because OrmLite doesn't generate FK relationships)
fix build.grade to use griddle mod
fix OrmLite test to delete everything after it finishes so the DB sizes are now all similar
fix Sprinkles test to pass in transaction to save call but it makes no different in Sprinkles' snail speed :-P

kenkyee commented:
Addressed all the PR comments I think?

agrosner commented:
@kenkyee make it mergeable and then it can be merged.

kenkyee committed:
Merge branch 'master' into feature/ken/ormlite

Conflicts:
	app/build.gradle

kenkyee committed:
fix possible activity leak by using grenade event bus; handle screen rotation

kenkyee commented:
fixed activity leak by using eventbus; also handled screen rotation
